File: src/main/java/net/clgd/ccemux/rendering/tror/AsyncScanner.java
package net.clgd.ccemux.rendering.tror;

import java.io.InputStream;
import java.util.ArrayList;
import java.util.Scanner;

class AsyncScanner implements Runnable {
	private final Scanner input;
	private ArrayList<String> lines = new ArrayList<>();

	public AsyncScanner(InputStream input) {
		this.input = new Scanner(input);
	}

	public boolean hasLines() {
		return lines.size() > 0;
	}

	public ArrayList<String> getLines() {
		ArrayList<String> out = new ArrayList<>(lines);
		lines.clear();
		return out;
	}

	public Thread start() {
		Thread t = new Thread(this);
		t.setDaemon(true);
		t.start();
		return t;
	}

	@Override
	public void run() {
		while (input.hasNextLine()) {
			lines.add(input.nextLine());
		}
	}
}

File: src/main/kotlin/net/clgd/ccemux/rendering/tror/AsyncScanner.kt
package net.clgd.ccemux.rendering.tror

import java.io.Closeable
import java.io.InputStream
import java.util.*

class AsyncScanner(inputStream: InputStream) : Runnable, Closeable {
	val input = Scanner(inputStream)
	val lines = ArrayList<String>()
	val thread = Thread(this)

	private var keepRunning = false

	init {
		thread.isDaemon = true
	}

	fun hasLines() = lines.isNotEmpty()

	fun getLinesList(): ArrayList<String> {
		@Suppress("UNCHECKED_CAST")
		val out = lines.clone() as ArrayList<String>
		lines.clear()
		return out
	}

	fun start() = thread.start()

	override fun run() {
		keepRunning = true

		while (keepRunning && input.hasNextLine()) {
			lines.add(input.nextLine())
		}

		input.close()
	}

	override fun close() {
		keepRunning = false
	}
}
